### ISSUE TITLE
receiver optimisation: possible to further short-circuit early recursion stop check?

### ISSUE THREAD

**Issue by mharvey-jt** [cvmfs/cvmfs#3571]
Consider the directory structure:
```
TOP
└── L1
    ├── 1
    ├── 2
    └── 3
```
with a nested catalog at each directory; and a transaction that adds `TOP/L1/4` with a lease path of `TOP/L1/4`.
`DiffRec` will make the early recursion stop check 

https://github.com/cvmfs/cvmfs/blob/47a223e1e0398d8b32ecab091c5b1155cd78540c/cvmfs/catalog_diff_tool_impl.h#L186-L196
for all directories `1,2,3`.  This is a relatively slow operation, and can be a bottleneck when `L1` has very many nested catalogs.

Would it be possible to further short-circuit this check based on whether the path being considered is in the lease path (IsReportablePath()==true?)

**Comment by jblomer** [comment]
I think that's a good idea! If I'm not mistaken, `TOP/L1/{1,2,3}` are even _ignored_ paths, not only unreportable paths. For ignored paths, we know that the recursion will stop and that we never report them. Looks like it is safe to immediately move on when `old_path` or `new_path` should be ignored (while still moving the iterators `i_from`, `i_to` accordingly).

**Comment by mharvey-jt** [comment]
Does #3577 do the job?

**Comment by mharvey-jt** [comment]
perhaps the check can even move outside of that clause, to line 185?

**Comment by jblomer** [comment]
I think it does the job.  Yes, it should be fine to move it up. And it is sufficient to check for `old_path` or `new_path` because we already know that they are identical.

We can also, I think, make a similar optimization for the `ReportAddition` and `ReportRemoval` branches. We then may want to remove the `IsIgnoredPath()` check at the beginning of the method.

**Comment by koallen** [comment]
> I think it does the job. Yes, it should be fine to move it up. And it is sufficient to check for `old_path` or `new_path` because we already know that they are identical.
> 
> We can also, I think, make a similar optimization for the `ReportAddition` and `ReportRemoval` branches. We then may want to remove the `IsIgnoredPath()` check at the beginning of the method.

I agree, the changes would essentially replace the check at the start of the function as an even early recursion stop. Provided it's implemented correctly (i.e. as you mentioned do it for `ReportAdditional` and `ReportRemoval` as well)
